text/0000-copy-clone-semantics.md

```diff
@@ -0,0 +1,66 @@
+- Feature Name: N/A
+- Start Date: 01 March, 2016
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+With specialization on the way, we need to talk about the semantics of
+`<T as Clone>::clone() where T: Copy`.
+
+It's generally been an unspoken rule of Rust that a `clone` of a `Copy` type is
+equivalent to a `memcpy` of that type; however, that fact is not documented
+anywhere. This fact should be in the documentation for the `Clone` trait, just
+like the fact that `T: Eq` should implement `a == b == c == a` rules.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, `Vec::clone()` is implemented by creating a new `Vec`, and then
+cloning all of the elements from one into the other. This is slow in debug mode,
+and may not always be optimized (although it often will be). Specialization
+would allow us to simply `memcpy` the values from the old `Vec` to the new
+`Vec` in the case of `T: Copy`. However, if we don't specify this, we will not
+be able to, and we will be stuck looping over every value.
```

Comment:
We will be stuck relying on the optimizer to figure it out. Which it can do and insert `memcpy`, typically for `Copy` values of 8 bytes or smaller, it looks like.